File: 20-2/HW4_Difference_Table.py
import numpy as np
def datainput(n,storage): #n개 데이터포인트를 storage에 입력받는 함수
    for i in range(0,n,1):
        #globals()['vy{}'.format(i)]=float(input('Enter Y{} : '.format(i))) #동적 변수명 할당방식
        x=float(input('Enter X{} : '.format(i)))
        storage[0,i]=x #1행에 x값 저장

    for i in range(0,n,1):
        # globals()['vy{}'.format(i)]=float(input('Enter Y{} : '.format(i))) #동적 변수명 할당방식
        y=float(input('Enter Y{} : '.format(i)))
        storage[1,i]=y  #2행에 y값 저장

def diff(n,storage): #차분계산
    for i in range(1,n,1): #차수에 대한 반복 i=1,2....n-1
        for j in range(0,n-1,1): #데이터수에 대한 반복 j=0,1,2....n-2
            storage[i+1,j]=storage[i,j+1]-storage[i,j]
            #storage배열 [2,0] [2,1] [2,2]...에 ([1,1]-[1,0), ([1,2]-[1,1])....([1,n-1]-[1,n-2])를 할당

def divdiff(n,storage): #분할차분계산
    for i in range(1,n,1): #차수에 대한 반복 i=1,2....n-1
        for j in range(0,n-1,1): #데이터수에 대한 반복 j=0,1,2....n-2
            storage[i+1,j]=(storage[i,j+1]-storage[i,j])/(storage[0,j+i]-storage[0,j])
            #1차 : x1-x0, x2-x1, x3-x2
            #2차 : x2-x0, x3-x1
            #3차 : x3-x0

ndata=int(input('Number of data points? : '))
storage=np.zeros((10,10))
datainput(ndata,storage)

print('X Datas : {}'.format(storage[0,0:ndata]))
print('Y Datas : {}'.format(storage[1,0:ndata]))

diff(ndata,storage)
for i in range(ndata-1):
    print('{} order differences : {}'.format(i+1,storage[i+2,0:ndata-(i+1)]))

divdiff(ndata,storage)
for i in range(ndata-1):
    print('{} order divided differences : {}'.format(i+1,storage[i+2,0:ndata-(i+1)]))

print('Stored array is : \n{}'.format(storage))

File: 20-2/C++/HW4_difference_table/HW4_difference_table/HW4_difference_table.cpp
#include<iostream>
using namespace std;
#define N 10

void datainput(int n,double arr[N][N]) //데이터 입력함수
{
	double x, y;
	for (int i = 0; i < n; i++) // x데이터 입력
	{
		cout << "Enter X" << i<<" : ";
		cin >> x;
		arr[0][i] = x;
	}
	for (int i = 0; i < n; i++) // y데이터 입력
	{
		cout << "Enter Y" << i << " : ";
		cin >> y;
		arr[1][i] = y;
	}
}

void diff(int n, double arr[N][N]) //차분계산 및 저장함수
{
	for (int i = 1; i < n; i++)
	{
		for (int j = 0; j < n - 1; j++)
		{
			arr[i + 1][j] = arr[i][j + 1] - arr[i][j];
		}
	}
}

void divdiff(int n, double arr[N][N]) //분할차분계산 및 저장함수
{
	for (int i = 1; i < n; i++)
	{
		for (int j = 0; j < n - 1; j++)
		{
			arr[i + 1][j] = (arr[i][j + 1] - arr[i][j]) / (arr[0][j + i] - arr[0][j]);
		}
	}
}

int main()
{
	int ndata;
	cout << "Number of data points? : ";
	cin >> ndata;
	double storage[N][N] = { 0 };
	datainput(ndata, storage);


	//입력된 X Y 데이터 출력
	cout << "Input X datas : ";
	for (int i = 0; i < ndata; i++)
	{
		cout << storage[0][i] << " ";
	}
	cout <<endl<< "Input Y datas : ";
	for (int i = 0; i < ndata; i++)
	{
		cout << storage[1][i] << " ";
	}
	cout << endl;


	//차분 계산 및 출력
	diff(ndata, storage);
	for (int i = 0; i < ndata - 1; i++)
	{
		cout << i + 1 << " order differences : ";
		for (int j = 0; j < ndata - (i + 1); j++)
		{
			cout << storage[i + 2][j] << " ";
		}
		cout << endl;
	}


	//분할차분 계산 및 출력
	divdiff(ndata, storage);
	for (int i = 0; i < ndata - 1; i++)
	{
		cout << i + 1 << " order divided differences : ";
		for (int j = 0; j < ndata - (i + 1); j++)
		{
			cout << storage[i + 2][j] << " ";
		}
		cout << endl;
	}
}
